### PR TITLE
Fixed lint lua errors in src

### PR DESCRIPTION
**Pull request by sameedali**
This pull request has commits which fix the various warnings and errors thrown by luacheck when run on the lua files inside the src directory.

### CODE REVIEW

File: src/nvim/generators/gen_api_dispatch.lua [hunk 1]
```diff
@@ -135,9 +135,9 @@ for i,f in ipairs(shallowcopy(functions)) do
 end
 
 -- don't expose internal attributes like "impl_name" in public metadata
-exported_attributes = {'name', 'parameters', 'return_type', 'method',
+local exported_attributes = {'name', 'parameters', 'return_type', 'method',
                        'since', 'deprecated_since'}
```
Comment (bfredl) [inline src/nvim/generators/gen_api_dispatch.lua]: need to adjust indentation of follow-up line.

File: src/nvim/generators/gen_api_dispatch.lua [hunk 2]
```diff
@@ -301,19 +305,19 @@ end
...
-mpack_output = io.open(mpack_outputf, 'wb')
+local mpack_output = io.open(mpack_outputf, 'wb')
 mpack_output:write(mpack.pack(functions))
 mpack_output:close()
 
-local function include_headers(output, headers)
-  for i = 1, #headers do
-    if headers[i]:sub(-12) ~= '.generated.h' then
-      output:write('\n#include "nvim/'..headers[i]..'"')
+local function include_headers(arg_output, arg_headers)
```
Comment (bfredl) [inline src/nvim/generators/gen_api_dispatch.lua]: maybe instead rename the global `output` ? At least I find no precedent for `arg_` .

Comment (sameedali) [inline src/nvim/generators/gen_api_dispatch.lua]: I renamed it to "output_handle" is that fine? I couldn't think of a more appropriate name for the global output so didnt change it. Also named the arg_headers to "headers_to_include"

File: src/nvim/generators/gen_declarations.lua [hunk 3]
```diff
@@ -225,14 +224,12 @@ while init ~= nil do
     else
       declline = declline - 1
     end
-  elseif init < declendpos then
-    -- Skipping over declaration
-  elseif is_needed_file then
+  elseif (not (init < declendpos)) and is_needed_file then
```
Comment (justinmk) [inline src/nvim/generators/gen_declarations.lua]: We lost documentation here (the deleted comment). Would be better to add an inline `-- luacheck: ...` ignore.

Comment (sameedali) [reply src/nvim/generators/gen_declarations.lua]: I have updated the code to reflect these changes.

File: src/nvim/generators/gen_unicode_tables.lua [hunk 4]
```diff
@@ -208,9 +208,7 @@ local build_width_table = function(ut_fp, dataprops, widthprops, widths,
       -- But use all chars from a range.
       local dp = dataprops[dataidx]
       if (n_last > n) or (not (({Mn=true, Mc=true, Me=true})[dp[3]])) then
-        if start >= 0 and end_ + 1 == n then
-          -- Continue with the same range.
-        else
+        if not (start >= 0 and end_ + 1 == n) then
```
Comment (justinmk) [inline src/nvim/generators/gen_unicode_tables.lua]: same here, the old code was intentional documentation. Also when adding `-- luacheck: ignore` it should only ignore the specific luacheck code number. In this case it's 542, see http://luacheck.readthedocs.io/en/stable/warnings.html

```lua
if start >= 0 and end_ + 1 == n then  -- luacheck: ignore 542
  -- Continue with the same range.
else
```

File: src/nvim/generators/gen_declarations.lua [hunk 5]
```diff
@@ -225,14 +224,14 @@ while init ~= nil do
     else
       declline = declline - 1
     end
-  elseif init < declendpos then
+  elseif init < declendpos then -- luacheck: ignore
```
Comment (justinmk) [inline src/nvim/generators/gen_declarations.lua]: -- luacheck: ignore 542

**Comment by XVilka** [comment]
@sameedali maybe a good time to rebase and solve conflicts?

**Comment by XVilka** [comment]
Apparently Travis fails to pass functional tests.

**Comment by blueyed** [comment]
@XVilka 
Needs a rebase first - worth trying if it passed then.

@sameedali 
Do you plan to pick this up again?

**Comment by justinmk** [comment]
I will take care of this next week.

File: src/nvim/lua/vim.lua [hunk 6]
```diff
@@ -38,8 +38,8 @@
 -- Returns:
 --    {errcode}, {output}
 local function _system(cmd)
-  local out = vim.api.nvim_call_function('system', { cmd })
-  local err = vim.api.nvim_get_vvar('shell_error')
+  local out = vim.api.nvim_call_function('system', { cmd }) --luacheck: ignore 113
+  local err = vim.api.nvim_get_vvar('shell_error') --luacheck: ignore 113
```
Comment (blueyed) [inline src/nvim/lua/vim.lua]: We should have a `.luacheckrc` instead, with:
```lua
-- Global objects defined by the C code
read_globals = {
    "vim",
}
```

**Comment by sameedali** [comment]
Hi,
Sorry about the delay. I just rebased it and and rechecked that all Lua files in `src` report no errors with luacheck. The tests are passing now. Once again sorry for the tardiness on my part.

File: src/nvim/lua/vim.lua (hunk 6 above)
Comment (blueyed) [reply src/nvim/lua/vim.lua]: See https://github.com/awesomewm/awesome/blob/3e3c34bdd194cf75cc16a87e82b11f8df3e7c5de/.luacheckrc for more ideas.

Comment (blueyed) [reply src/nvim/lua/vim.lua]: Also please consider using `[skip appveyor]` with following commits (since linting is run on Travis).

**Comment by blueyed** [comment]
Also this appears to have been rebased/merged wrongly in some way.

**Comment by blueyed** [comment]
> Once again sorry for the tardiness on my part.

No worries.  After all Justin said to take over anyway..?

This should also run `luacheck` then with the lint job I assume.

File: src/nvim/options.lua [hunk 7]
```diff
@@ -188,7 +188,7 @@ return {
     },
     {
       full_name='belloff', abbreviation='bo',
-      deny_duplicates=true,
+      deny_duplicates=true, -- luacheck: ignore 314
       type='string', list='comma', scope={'global'},
       deny_duplicates=true,
```
Comment (blueyed) [inline src/nvim/options.lua]: That's an actual error that should not be ignored.

Comment (sameedali) [reply src/nvim/options.lua]: Thanks. Fixed it

File: src/nvim/lua/vim.lua (hunk 6 above)
Comment (sameedali) [reply src/nvim/lua/vim.lua]: Okay, I have added the .luacheck rc in src/ and removed the luacheck ignore flags for vim global var in this file.

**Comment by sameedali** [comment]
> Also this appears to have been rebased/merged wrongly in some way.

Yeah, i think I made a mistake and force pushed once. I don't know how to correct that. Should I redo all changes in one commit and generate another pull request? I'm a little confused as to how should I fix this?

**Comment by blueyed** [comment]
Force-pushing is fine, but it looks like you've rebased to much..
Try `git rebase -i origin/master` and then select/keep/pick only your commits.

File: src/nvim/lua/vim.lua (hunk 6 above)
Comment (blueyed) [reply src/nvim/lua/vim.lua]: For future commits, please use `[skip appveyor]` with the commit subject (first line).

**Comment by blueyed** [comment]
@sameedali 
Also just for information: it is usually a good idea to use a separate Git branch for PRs, but this cannot be changed now - just remember it for future submissions.

**Comment by sameedali** [comment]
> Force-pushing is fine, but it looks like you've rebased to much..
> Try `git rebase -i origin/master` and then select/keep/pick only your commits.

@blueyed I just rebased it. I think it is correct now

**Comment by blueyed** [comment]
@sameedali 
Looks better, thanks!

As for cmake/make targets:
In general I think we should have a `lint-lua` target then, which gets called/used by `lint` then also, and the existing `lint` might be renamed/factored out to `lint-clint` then.
But let's wait for @justinmk / others for this.

File: src/nvim/lua/vim.lua [hunk 8]
```diff
@@ -38,8 +38,8 @@
 -- Returns:
 --    {errcode}, {output}
 local function _system(cmd)
-  local out = vim.api.nvim_call_function('system', { cmd })
-  local err = vim.api.nvim_get_vvar('shell_error')
+  local out = vim.api.nvim_call_function('system', { cmd }) --luacheck: ignore 113
```
Comment (bfredl) [inline src/nvim/lua/vim.lua]: can these be removed, now that the luacheckrc is added?

Comment (blueyed) [reply src/nvim/lua/vim.lua]: Yes.

Comment (sameedali) [reply src/nvim/lua/vim.lua]: I just removed them.

**Comment by blueyed** [comment]
Note/JFI: there's `test/.luacheckrc` already.

**Comment by bfredl** [comment]
LGTM, is this ready? Perhaps need a rebase to check with recent lua changes.

**Review by justinmk — APPROVED** [review]

**Comment by blueyed** [comment]
Finishing this in https://github.com/neovim/neovim/pull/10414.

**Pull request closed without merging**